src/spikeinterface/core/generate.py

```diff
@@ -220,16 +221,340 @@ def add_synchrony_to_sorting(sorting, sync_event_ratio=0.3, seed=None):
-    spikes_all = spikes_all[sort_idxs]
+    sort_idxs = np.lexsort([spikes_duplicated["sample_index"], spikes_duplicated["segment_index"]])
+    spikes_duplicated = spikes_duplicated[sort_idxs]
 
-    sorting = NumpySorting(spikes=spikes_all, sampling_frequency=sorting.sampling_frequency, unit_ids=unit_ids)
+    synchronous_spikes = NumpySorting(spikes_duplicated, sorting.get_sampling_frequency(), unit_ids)
+    sorting = TransformSorting.add_from_sorting(sorting, synchronous_spikes)
 
     return sorting
 
 
+def generate_sorting_to_inject(
+    sorting: BaseSorting,
+    num_samples: List[int],
+    max_injected_per_unit: int = 1000,
+    injected_rate: float = 0.05,
+    refractory_period_ms: float = 1.5,
+    seed=None,
+) -> NumpySorting:
+    """
+    Generates a sorting with spikes that are can be injected into the already existing sorting without violating
+    the refractory period.
+
+    Parameters
+    ----------
+    sorting : BaseSorting
+        The sorting object
+    num_samples: list of size num_segments.
+        The number of samples in all the segments of the sorting, to generate spike times
+        covering entire the entire duration of the segments
+    max_injected_per_unit: int, default 1000
+        The maximal number of spikes injected per units
+    injected_rate: float, default 0.05
+        The rate at which spikes are injected
+    refractory_period_ms: float, default 1.5
+        The refractory period that should not be violated while injecting new spikes
+    seed: int, default None
+        The random seed
+
+    Returns
+    -------
+    sorting : NumpySorting
+        The sorting object with the spikes to inject
+
+    """
+
+    injected_spike_trains = [{} for seg_index in range(sorting.get_num_segments())]
+    t_r = int(round(refractory_period_ms * sorting.get_sampling_frequency() * 1e-3))
+
+    rng = np.random.default_rng(seed=seed)
+
+    for segment_index in range(sorting.get_num_segments()):
+        for unit_id in sorting.unit_ids:
+            spike_train = sorting.get_unit_spike_train(unit_id, segment_index=segment_index)
+            n_injection = min(max_injected_per_unit, int(round(injected_rate * len(spike_train))))
+            # Inject more, then take out all that violate the refractory period.
+            n = int(n_injection + 10 * np.sqrt(n_injection))
+            injected_spike_train = np.sort(
+                np.random.uniform(low=0, high=num_samples[segment_index], size=n).astype(np.int64)
+            )
+
+            # Remove spikes that are in the refractory period.
+            violations = np.where(np.diff(injected_spike_train) < t_r)[0]
+            injected_spike_train = np.delete(injected_spike_train, violations)
+
+            # Remove spikes that violate the refractory period of the real spikes.
+            # TODO: Need a better & faster way than this.
+            min_diff = np.min(np.abs(injected_spike_train[:, None] - spike_train[None, :]), axis=1)
```

Comment:
This is an enormous matrix!!!!
We should have a better way with this kind of approach:
https://github.com/SpikeInterface/spikeinterface/blob/872da6ef16b1c0113da28d62c00dc74fbd5eb556/src/spikeinterface/postprocessing/correlograms.py#L237

In short make a sorted conatenated vector with spikes times + spike labels ordered.
And then with a small on shift we can detect and count how many spikesare violating.

(More or less the old algo for matching even)

What do you think ?

Reply:
I did not touch this function. I guess the modification of this (and I agree this is not optimal) is out of the scope of this PR